src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,411 @@
+        //
+        //    <if-statement-sequence>
+        //        : if (<section-expr>) { <unreachable-end-point> }, <if-statement-sequence>
+        //        : if (<section-expr>) { <unreachable-end-point> }, ( return | throw )
+        //        | <if-statement>
+        //
+        //    <if-statement>
+        //        : if (<section-expr>) { _ } else <if-statement>
+        //        | if (<section-expr>) { _ } else { _ }
+        //        | if (<section-expr>) { _ }
+        //        | { <if-statement-sequence> }
+        //
+        //    <section-expr>
+        //        : <section-expr> || <pattern-expr>
+        //        | <pattern-expr>
+        //
+        //    <pattern-expr>
+        //        : <pattern-expr> && <expr>                         // C#
+        //        | <expr0> is <pattern>                             // C#
+        //        | <expr0> is <type>                                // C#
+        //        | <expr0> == <const-expr>                          // C#, VB
+        //        | <expr0> <comparison-op> <const>                  //     VB
+        //        | ( <expr0> >= <const> | <const> <= <expr0> )
+        //           && ( <expr0> <= <const> | <const> >= <expr0> )  //     VB
+        //        | ( <expr0> <= <const> | <const> >= <expr0> )
+        //           && ( <expr0> >= <const> | <const> <= <expr0> )  //     VB
+        //
+        internal abstract class Analyzer
+        {
+            public abstract bool CanConvert(IConditionalOperation operation);
+            public abstract bool HasUnreachableEndPoint(IOperation operation);
+
+            // Holds the expression determined to be used as the target expression of the switch
+            private SyntaxNode? _switchTargetExpression;
+            private readonly ISyntaxFactsService _syntaxFacts;
+            private readonly Feature _features;
+
+            protected Analyzer(ISyntaxFactsService syntaxFacts, Feature features)
+            {
+                _syntaxFacts = syntaxFacts;
+                _features = features;
+            }
+
+            public bool Supports(Feature feature)
+                => (_features & feature) != 0;
+
+            public (ImmutableArray<AnalyzedSwitchSection>, SyntaxNode) AnalyzeIfStatementSequence(ReadOnlySpan<IOperation> operations)
+            {
+                var sections = ArrayBuilder<AnalyzedSwitchSection>.GetInstance();
+                if (!ParseIfStatementSequence(operations, sections, out var defaultBodyOpt))
+                {
+                    sections.Free();
+                    return default;
+                }
+
+                if (defaultBodyOpt is object)
+                {
+                    sections.Add(new AnalyzedSwitchSection(labels: default, defaultBodyOpt, defaultBodyOpt.Syntax));
+                }
+
+                return (sections.ToImmutableAndFree(), _switchTargetExpression!);
```

Comment:
_switchTargetExpression is initially unset be we don't want to allow assigning null values, plus, at the end it is expected to be set.

I think `Expression _switchTargetExpression = null!` fits the above requirements?

But we need to assert anyways though I think it should be a DEBUG assertion - we don't have access to RoslynDebug added in https://github.com/dotnet/roslyn/pull/38646, should we add one for Workspaces project?

Reply:
You can either RoslynDebug.Assert() if you want or you can use Contract.ThrowIfNull(). I have no problem having a release assert in that I'm not entirely confident this would not crash if this was null...

Reply:
@sharwell also has a nice way you can annotate this as "it might be null to start but can't be assigned null" but I forget what it is off hand...